eth/downloader/downloader.go

```diff
@@ -382,207 +396,274 @@ func (d *Downloader) Terminate() {
+		for arrived := false; !arrived; {
+			select {
+			case <-d.cancelCh:
+				return 0, errCancelHashFetch
+
+			case hashPack := <-d.hashCh:
+				// Discard anything not from the origin peer
+				if hashPack.peerId != p.id {
+					glog.V(logger.Debug).Infof("Received hashes from incorrect peer(%s)", hashPack.peerId)
 					break
 				}
-			}
-			// Insert all the new hashes, but only continue if got something useful
-			inserts := d.queue.Insert(hashPack.hashes, false)
-			if len(inserts) == 0 && !done {
-				glog.V(logger.Debug).Infof("Peer (%s) responded with stale hashes", active.id)
-				return errBadPeer
-			}
-			if !done {
-				// Check that the peer is not stalling the sync
-				if len(inserts) < MinHashFetch {
-					return errStallingPeer
+				// Make sure the peer actually gave something valid
+				hashes := hashPack.hashes
+				if len(hashes) != 1 {
+					glog.V(logger.Debug).Infof("%v: invalid search hash set (%d)", p, len(hashes))
+					return 0, errBadPeer
+				}
+				arrived = true
+
+				// Modify the search interval based on the response
+				block := d.getBlock(hashes[0])
+				if block == nil {
+					end = check
+					break
 				}
-				// Try and fetch a random block to verify the hash batch
-				// Skip the last hash as the cross check races with the next hash fetch
-				cross := rand.Intn(len(inserts) - 1)
-				origin, parent := inserts[cross], inserts[cross+1]
-				glog.V(logger.Detail).Infof("Cross checking (%s) with %x/%x", active.id, origin, parent)
-
-				d.checks[origin] = &crossCheck{
-					expire: time.Now().Add(blockSoftTTL),
-					parent: parent,
+				if block.NumberU64() != check {
+					glog.V(logger.Debug).Infof("%v: non requested hash #%d [%x…], instead of #%d", p, block.NumberU64(), block.Hash().Bytes()[:4], check)
+					return 0, errBadPeer
 				}
-				go active.getBlocks([]common.Hash{origin})
+				start = check
 
-				// Also fetch a fresh batch of hashes
-				getHashes(head)
-				continue
-			}
-			// We're done, prepare the download cache and proceed pulling the blocks
-			offset := uint64(0)
-			if block := d.getBlock(head); block != nil {
-				offset = block.NumberU64() + 1
-			}
-			d.queue.Prepare(offset)
-			finished = true
+			case <-d.blockCh:
+				// Out of bounds blocks received, ignore them
 
-		case blockPack := <-d.blockCh:
-			// Cross check the block with the random verifications
-			if blockPack.peerId != active.id || len(blockPack.blocks) != 1 {
-				continue
+			case <-d.headerCh:
+				// Out of bounds eth/62 block headers received, ignore them
+
+			case <-d.bodyCh:
+				// Out of bounds eth/62 block bodies received, ignore them
+
+			case <-timeout:
+				glog.V(logger.Debug).Infof("%v: search hash timeout", p)
+				return 0, errTimeout
 			}
-			block := blockPack.blocks[0]
-			if check, ok := d.checks[block.Hash()]; ok {
-				if block.ParentHash() != check.parent {
-					return errCrossCheckFailed
-				}
-				delete(d.checks, block.Hash())
+		}
```

Comment:
The reason I duplicated all code belonging to eth61 vs 62 is because at a certain point I'd like to see eth/61 go, and then we could simply delete the corresponding functions and not have to worry about interweaved logic. I agree with generalizing things, but only as long as it doesn't cross the eth/61 and eth/62 boundaries so the latter can evolve without having potentially bad side effects on the former.